### ISSUE TITLE
New request: shamela.ws المكتبة الشاملة

### ISSUE THREAD

**Issue by RavanJAltaie** [openzim/zim-requests#1172]
*Please use the following format for a ZIM creation request (and delete unnecessary information)*

- Website URL: https://shamela.ws/
- License: Open free content
- Desired ZIM Title: المكتبة الشاملة
- Desired ZIM Description: مشروع يهدف لجمع ما يحتاجه طالب العلم من كتب وبحوث
- Desired ZIM Icon –png (URL or attach one): 
- Language (ISO 639-3): ara
- Is this a MediaWiki?: no

**Comment by RavanJAltaie** [comment]
Recipe created
https://farm.openzim.org/recipes/shamela.ws_ar_all
I'll update the library link once ready.
I already sent them an email to double check if any of the books in the library has a copyright.

**Comment by RavanJAltaie** [comment]
We've received an answer from the team that all the books in the website are more than 100 years old. In 20 years (their operation time) they've received only 2 claims of books copyrights and they've deleted the books immediately as per their website policy.

**Comment by hamoudak** [comment]
is that means it won't get crawled?! I have requested a zim file related to this website here[#986] I think it's public domain.
could this be made?

**Comment by Popolechien** [comment]
@hamoudak no no, this means we're good. You can follow the task on the link given above.

**Comment by hamoudak** [comment]
thank you, it's an valuable website for reading and studying.

**Comment by benoit74** [comment]
After 3 days, crawler progress is 3% (100753 / 2859505). 2.8 million links to explore is way too much.
I cancelled the task and disabled the recipe. We need to find another way of ZIMing this website, this is not feasible with zimit, at least as-is.

**Comment by hamoudak** [comment]
@benoit74 could my request [#986] be created; its one of five archives related to this domain, or I have to wait for some reason.

**Comment by benoit74** [comment]
The idea of dividing the ZIM per category as on the website is a good one.

And looking a bit more into it, I don't get why we ended-up with 2M links.

Anyway, I've started a first sub-recipe of category 34: https://farm.openzim.org/recipes/shamela.ws_ar_34

In this new recipe, ZIM name, title and description are very bad, this will have to be fixed, but at least let's see how it goes.

**Comment by hamoudak** [comment]
I can give you the names of these categories in arabic ; I know arabic very well.

**Comment by benoit74** [comment]
> why it ended with millions of pages; beacuse there are some books has many volumes like dictionaries, literature, explanations and interpreting quran. in addition to this there is a category called : sunna books (category/6) , most of this one has many links due to each page may have just two lines of text or one for a big book, so it ends eventually with many links.

OK, so what I did for `[ al-shir-wa-dawawinu] peotry diwans` is not going to work for all categories. I basically asked to explore only links listed in the category page, so if I understand you well, it will explore only the books but not their volumes.

If I get you correctly, what we would like is tell the scraper to:
- find all books URLs listed on the category page
- for all these URLs, explore the book URL and all its sub-URLs
- also explore all authors URLs since it (it might probably contain some "external links" because some authors will probably have books in another category which will hence not end inside the ZIM).

For instance, for https://shamela.ws/category/4, we want the book https://shamela.ws/book/23622 but also https://shamela.ws/book/23622/1, https://shamela.ws/book/23622/2 and so on, and also https://shamela.ws/author/263 ; and so on for all other books of the category.

Is this correct? Do we have other links / pages which would be needed in each ZIM per category?

All that being said, I don't know yet how to do it with zimit, but at least it is important to understand what we would like to achieve ^^

**Comment by benoit74** [comment]
> I can give you the names of these categories in arabic ; I know arabic very well.
this category called: [ al-shir-wa-dawawinu] peotry diwans.
arabic : الشعر ودواوينه

Glad you can help on this, thank a lot. Once we have a working plan, I will come back to you about what we need precisely.

**Comment by hamoudak** [comment]
first it will only explore the links but not their sub-pages. and you are absolutely right in all the three points you gave. I have made more than 200 (highly important) books of this domain with youzimit , when I did a basic crawl. I got just the titles (the contents of the book) not the sub-pages. so I went to the custom scope and give it the right parameters to the sub-pages links. I got it work then.


here;s the zim file of a book I made.
https://archive.org/download/sirat-ibn-hisham_904a56f0
[sirat-ibn-hisham_log.txt](https://github.com/user-attachments/files/17269013/sirat-ibn-hisham_log.txt)

![Image](https://github.com/user-attachments/assets/567a6529-80ef-490d-aa39-550bb3f78a1f)

**Comment by benoit74** [comment]
I think I've achieved to build a pretty good ZIM of category 34. You can see preview at https://dev.library.kiwix.org/#lang=&q=34 (this is not the final URL, and never guaranteed to work, this is just dev server).

I'm currently running again the recipe to update the icon (which is blank) and to update the CSS of HTML pages inside the ZIM (to hide useless things when offline).
What do you think?

My main concern is that it took 7 hours, which is not that bad given the scraper had to explore 7966 links, which gives us an average of 3 secs per link, but this was for only 25 books. I don't know what this will mean for a huge category like category 6 with 1227 books and very huge ones like https://shamela.ws/book/13174. 

For the new task which is currently processing, I've increased the number of parallel worker to 4, let's hope it will not trigger something bad on the upstream server.

I will also retrieve all main books page to count the number of links per category and have an estimate of total time per ZIM.

Do you have any idea of how often we should update the ZIM? E.g. how often are they adding new books, or how many books are they adding per month / quarter / year?

Since it looks like we will finally have a plan, it is now time to ask for your help regarding ZIM metadata.

For every ZIM (and hence category for now), we will need:
- a `selection`: this is what will go into the ZIM name, which will be named `shamela.wa_ar_<selection>` (without the `<` and `>`). Since we are doing one ZIM per category, the `selection` should be more or less the category. It should be as short as possible, but also as expressive as possible. It can contain only alphanumeric characters and the dash. If I understand you well, I imagine that for category 34 it should be `34-al-shir-wa-dawawinu` (I'm not sure adding the category number will help ... it would help us to maintain the ZIM at least ^^)
- a `title`: this is the ZIM title, displayed in all readers. It is limited to 30 characters. It should help to identify which ZIM the user is going to open. It must be in ZIM native language, so arabic, and user-friendly (i.e. it is not used by machines). If it was in English, I would for instance consider to use `shamela.ws books: category 34` since it is difficult to be more expressive in only 30 characters. I don't really like it to be honest, it is a bit ugly, but at the same time I don't find how to fit more in 30 chars. Maybe `shamela poetry diwans`? Not sure it will be possible for all categories, and it is less precise than the first alternative I proposed
- a `description`: this is the ZIM description, displayed in all readers. It is limited to 80 characters. It should help understand what is inside the ZIM, as a complement to the title. It must be in ZIM native language, so arabic, and user-friendly (i.e. it is not used by machines). If it was in English and I understand you well, I would probably use something like `Books of shamela.ws collection, category 34 of poetry diwans` (maybe adding something about what these books are would be interesting, are they about art, religion, daily life, law, mechanics, technology, ... didn't understood this so far)

Could you propose something for category 34 first? Please do not hesitate to ask friends for feedback as well on these, it is hard work and often good ideas might come from interactions with others.

**Comment by hamoudak** [comment]
you actually made it very good and a complete one. I 've download it from the farm before you post this comment. everything work as intended. for the things you'll hide it, I don't know much a bout it but the zim file as it is shown on the website is good enough. 

I  have noticed for a long time being on the website they adding a few books; about 7 books in a six weeks , their uploading new books is random, one time I just saw two new books after nearly two months. not much they add but working hard to add new ones.

now I think if you leave categories without numbers will be better. their names would define that, am going to work on them tomorrow; I'll update you with anything I see. Its my pleasure to help as much as I could with you.

**Comment by benoit74** [comment]
> I have noticed for a long time being on the website they adding a few books; about 7 books in a six weeks , their uploading new books is random, one time I just saw two new books after nearly two months. not much they add but working hard to add new ones.

OK, thank you. Updating the ZIM once per semester is hence probably a bare minimum. But I would like to avoid creating multiple ZIM in parallel to avoid overwhelming their server, so I doubt we can update once per quarter or it would mean mostly always having a ZIM update running.

> now I think if you leave categories without numbers will be better. their names would define that, am going to work on the "metadata" tomorrow; I'll update you with anything I see. Its my pleasure to help you as much as I could .

Thank you !

> note: I see the first zim is to be better than the updated one; its not vital like the original and I hope to keep it as it was for offline use.

I'm not sure I get you here. Do you mean that the links I hide in latest version made the ZIM worse? I'm very interested to have your feedback on this, we always considered that it is better to hide as many "broken" external links since they usually do not work in offline scenario and we've considered so far they will bring only frustration. 

> I will also retrieve all main books page to count the number of links per category and have an estimate of total time per ZIM.

On this estimate, as expected category 6 is the biggest one with approximately 1.1 million links. Given the speed of last task with 4 parallel workers, it should complete in about 10 days which is OK.

**Comment by hamoudak** [comment]
> > I have noticed for a long time being on the website they adding a few books; about 7 books in a six weeks , their uploading new books is random, one time I just saw two new books after nearly two months. not much they add but working hard to add new ones.
> 
> OK, thank you. Updating the ZIM once per semester is hence probably a bare minimum. But I would like to avoid creating multiple ZIM in parallel to avoid overwhelming their server, so I doubt we can update once per quarter or it would mean mostly always having a ZIM update running.
> 
> > now I think if you leave categories without numbers will be better. their names would define that, am going to work on the "metadata" tomorrow; I'll update you with anything I see. Its my pleasure to help you as much as I could .
> 
> Thank you !
> 
> > note: I see the first zim is to be better than the updated one; its not vital like the original and I hope to keep it as it was for offline use.
> 
> I'm not sure I get you here. Do you mean that the links I hide in latest version made the ZIM worse? I'm very interested to have your feedback on this, we always considered that it is better to hide as many "broken" external links since they usually do not work in offline scenario and we've considered so far they will bring only frustration.
> 
> > I will also retrieve all main books page to count the number of links per category and have an estimate of total time per ZIM.
> 
> On this estimate, as expected category 6 is the biggest one with approximately 1.1 million links. Given the speed of last task with 4 parallel workers, it should complete in about 10 days which is OK.

**Comment by benoit74** [comment]
> yes I see the latest changes on cat 34 is not suitable for reading books, the white theme make it un comfortable to the eye for reading anything

I did not intentionally changed anything regarding themes, I just removed external links. And I always had a white theme, with a toggle (which is still there) to enable dark theme. So I don't get your point. Can you share a screenshot of previous ZIM and new one, or small videos?

**Comment by hamoudak** [comment]
there's a simple issue (related to theme and links) in the old zim you've created when am browsing it shows the links and themes as it is online then when i go forward or back  while browsing a book it disppears completely ! 
this is : shamela.ws_ar_34_2024-10 _size 24.4MG   page:80
![Image](https://github.com/user-attachments/assets/395b6be1-4e00-4ae8-87df-95f33b92a7e2)

page:81 for the same file.

![Image](https://github.com/user-attachments/assets/13ba399b-36e8-4f46-963e-aa0d11b75007)

for the theme you removed it was the light blue one, and you're right no difference between the two zims. so it's the light blue; could you keep it, please :)

**Comment by benoit74** [comment]
I'm sorry but I still don't get it. What do you want to keep? The light blue header on the top of the ZIM? Isn't this just a zone with a link to contribute to shamela.ws, which is not going to work offline? Do you mean that I should keep the blue zone and just remove the link instead? (I feel like this blue zone was useless if empty, but I can easily add it back if you feel like it is useful even if empty)

**Comment by hamoudak** [comment]
yes; you got my point, it will relieve tired eyes a little bit when you reading a long time . you'll see a big difference.

 I almost done of the descriptions , also I have translated all category names into english ; it was the first time to translate religious idioms, so It had much time to be sure of the translated text. besides reading and searching related context articles.

**Comment by benoit74** [comment]
> edit: there's coloured texts, I think it will fit with.

do you mean the text in yellow leading to a contribute page? I can keep it as well, no strong opinion other than this link will be outside the ZIM, needing an online connection.

**Comment by hamoudak** [comment]
no I didn't mean that; I meant the texts within the book, its coloured with many colours for study and remembering. so the light blue will fit with those coloured words, but I don't mind if you keep the yellow that leads to contribute page. 

I  think I have finished of the metadata file , you can now have a look, and tell me what do you think or if anything should be edited etc.
[shamela_metadata_2024-10.txt](https://github.com/user-attachments/files/17296395/shamela_metadata_2024-10.txt)

**Comment by benoit74** [comment]
> no I didn't mean that; I meant the texts within the book, its coloured with many colours for study and remembering. so the light blue will fit with those coloured words. remove that yellow.

OK, I just relaunched category 34 with that change (and fixed ZIM name, title and description). Note that recipe is now at https://farm.openzim.org/recipes/shamela.ws_ar_alshir-wa-dawawinu for consistency with ZIM name.

> I think I have finished of the metadata file , you can now have a look, and tell me what do you think or if anything should be edited etc.

Thank you a lot for this metadata file! It looks ok at first sight. I might have few more questions as I dive into details category per category, but it is sufficient to get me started.

**Comment by hamoudak** [comment]
I think with the light blue it had become very good; please work on the updated metadata a bove; I fixed some typo.

**Comment by benoit74** [comment]
@Popolechien : ZIM for category 34 is ready to move to prod: https://dev.library.kiwix.org/#q=%D8%A7%D9%84%D9%85%D8%AC%D9%85%D9%88%D8%B9%D8%A9 ; can you please have a look before I publish it?

**Comment by Popolechien** [comment]
@benoit74 Not 100% but I suspect there is a formatting issue with that black bar over here: https://dev.library.kiwix.org/viewer#shamela.ws_ar_alshir-wa-dawawinu_2024-10/shamela.ws/author/3009
![Image](https://github.com/user-attachments/assets/3f2b2c27-1cfd-4246-81e8-aa66c53c136b)

the top link also redirect to shamela.ws, which is blocked as it is considered external link - any chance we can have it redirect to the zim's home page?

**Comment by hamoudak** [comment]
@Popolechien that's will be good, to be redirected to the home page.
before this zim go to prod, one little thing. there is this sign ( ; ) for seperating sentences, its for english not arabic. in arabic must be ( ؛ )
i've edited them all here [shamela_metadata_2024-10.txt](https://github.com/user-attachments/files/17309273/shamela_metadata_2024-10.txt)

so this zim metadata is:
the arabic with english letters: alshir-wa-dawawinu
title: دواوين الشعر؛ المجموعة رقم 34
description: دواوين الشعر العربي في الجاهلية وصدر الإسلام، وبعض الشروحات عليها

sorry to mention this  but now everything is complete from my side.

**Comment by benoit74** [comment]
OK, sorry for the broken layout, I broke thing with last CSS, I've fixed it, running again the recipe to update the ZIM ATM

> the top link also redirect to shamela.ws, which is blocked as it is considered external link - any chance we can have it redirect to the zim's home page?

Nope

> i've edited them all here

OK, I'm using your last file now.

> Edit: I have finished of category 4 urls for custom scope [category_4.txt](https://github.com/user-attachments/files/17311993/category_4.txt)

Thank you, but I already have built the list for all books per categories with a script, no need to do that for other categories manually.

**Comment by hamoudak** [comment]
:) that's cool

**Comment by benoit74** [comment]
>  its working fine now on "kiwix js pwa". thank you.

Thanks, new ZIM seems indeed ok now. @Popolechien anymore remark or shall I move this ZIM to prod?

**Comment by Popolechien** [comment]
Can you do a bit of CSS to remove/hide the search box?
![Image](https://github.com/user-attachments/assets/749abfab-c415-4938-bf0d-205c36085819)

Other than that and if @hamoudak has no other comment I think we'll be good to go.

**Comment by hamoudak** [comment]
"authors index" I think it needs a single zim file !

**Comment by hamoudak** [comment]
in config:
title | العقيدة؛ المجموعة رق 1
correction:
title | العقيدة؛ المجموعة رقم 1

رقم (not رق ) means: number

**Comment by benoit74** [comment]
Thank you... I just copy-pasted, so don't get how I achieved to make it wrong ...

**Comment by hamoudak** [comment]
don't mention it, it happens. maybe your mind was busy with something else. what matters is your work eventually to be perfect.

**Comment by benoit74** [comment]
I just realized that some titles (category 3 and 5 at least) have too long titles. It should fit within 30 chars. @hamoudak can you please suggest fixes?

**Comment by benoit74** [comment]
Yes, spaces count. And I'm speaking about the arabic titles and descriptions. The "name" in english is not really limited (but the shorter the better obviously since it will be placed into ZIM filename)

**Comment by hamoudak** [comment]
I have made descriptions a little bit shorter here; you can work with it:
[fixes-and-suggestions-discriptions.txt](https://github.com/user-attachments/files/17398016/fixes-and-suggestions-discriptions.txt)

**Comment by hamoudak** [comment]
why in https://dev.library.kiwix.org/#lang=&q=%D8%A7%D9%84%D8%B9%D9%82%D9%8A%D8%AF%D8%A9
is there رق  which means number is not updated as you did in its config .

**Comment by benoit74** [comment]
> why in https://dev.library.kiwix.org/#lang=&q=%D8%A7%D9%84%D8%B9%D9%82%D9%8A%D8%AF%D8%A9
is there رق which means number is not updated as you did in its config .

This is because the task was already running with "bad" description, as can be seen at https://farm.openzim.org/pipeline/c10047a3-8243-4e4b-b603-519d75d42d4c/debug, when I fixed it. I didn't wanted to interrupt it since it takes time to complete. I will run it again with the updated config, it is going to be fast, I've configured the task to keep an intermediate archive to rebuild the ZIM fast.

Thank you for your attention, much appreciated!

> and by the way I can tell you many mistakes in translation to arabic or "typo" for all the zim files on lib.kiwix.
such as the description of مدرسه (madrasa) . there is an arabic word is wrong فيدوات , it should be فيديوهات.
the plural of the word "video" .
and many others related to the discriptions of arabic wikipedia.

To fix this, you can open a dedicated issue "fix typo in arabic words" and precisely list everything you can. Unfortunately most of us are kiwix are not arabic writers / speakers at all, and we somehow trusted the person who worked with us to not input typos. We should probably double or triple check this.

**Comment by hamoudak** [comment]
yes, I know  the most  of you from Switzerland its hard to know arabic well. I was graduated from the faculty of languages and translation long time ago, but I was affected by some troubles. I do my best though. and I have the facility to revise texts.

would I open it in zim-requests?

**Comment by benoit74** [comment]
> would I open it in zim-requests?

Yes please

**Comment by hamoudak** [comment]
I think their server is temporarily down for maintenance. could you make my book request  till it comes back.

**Comment by benoit74** [comment]
[shamela.ws_ar_al-tafsir-3](https://farm.openzim.org/recipes/shamela.ws_ar_al-tafsir-3) failed while it was close to completion. It ran way slower than expected, I will have to check all this before starting again a task.

**Comment by hamoudak** [comment]
I have noticed that the minute it was done, and I wanted to ask you why it failed.  you may run a smaller cat now till you check this one. I see that maybe you made more tasks besides this one or it has nothing to do with it !

**Comment by hamoudak** [comment]
why scroll bar is on the left while its on the right side on the website and the desktop app version !

**Comment by benoit74** [comment]
I increased the number of worker of [shamela.ws_ar_al-tafsir-3](https://farm.openzim.org/recipes/shamela.ws_ar_al-tafsir-3) to 8 (instead of 4), let's see

**Comment by hamoudak** [comment]
I hope it ends good.

**Comment by hamoudak** [comment]
It has failed once again.  to let you know this cat 3 and and cat 6 has books that include too many pages loading slowly when you browsing online. so lowering the workers may be your best approach , for example to 2 workers or 1.

**Comment by hamoudak** [comment]
before you go scrapig https://farm.openzim.org/recipes/shamela.ws_ar_al-tajwid-wa-al-qiraat-5/config .
please remove this "J" letter in description.

**Comment by benoit74** [comment]
Done, thank you!

**Comment by hamoudak** [comment]
do you know that you have many missing pages in shamela.ws_ar_al-adab-32 ?

**Comment by benoit74** [comment]
Thank you, I didn't had time to review this in details unfortunately.

Since scrape took 10 days, I appreciate a lot your recommendation but this is however not really practicable in our context.

I will pause the recipes for now, we need to find a solution.

**Comment by hamoudak** [comment]
If there any moment that RavanJAltaie can talk to the domain owner to solve this issue or advice him to a solution to his server issues, that will help a lot. am sure he will welcome the idea of archiving his website, and to give him a version of these zim files to try, I think he doesn't know what all about. he himself has a related domain has archives of other forums. 

then. Ilya Kreymer can help with testing this domain issues, he's able to sovle this. I talked to him before about issues happen when am archiving an arabic domain with archive web.page extension. it was crashing , he solved it.

**Comment by Popolechien** [comment]
@hamoudak Sadly Ravan is not with Kiwix anymore. Feel free to ping the website owner, however, and cc hello@kiwix.org so that we can keep track of his approval (you can ping him in Arabic, this is just for record-keeping)

**Comment by hamoudak** [comment]
@benoit74 go right now for scraping, the server was fixed, no issues all day.  try to start with middle categories not big ones  to test this. i'll talk to you later about something in logs. but  everything is fine.
@Popolechien I was going to send a message after some time, but I found everything was fixed.

**Comment by benoit74** [comment]
Done, I've started few recipes

**Comment by benoit74** [comment]
A first ZIM came out: https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-ansab_2025-03/ ; looks very good as far as I've tested, @Popolechien @hamoudak can you test as well and confirm it is OK to be published on library ?

**Comment by hamoudak** [comment]
yes, I have tested everything once I downloaded the file yesterday. works very well to be published. I don't know much about the change here in the log file like in : 

Skipping item at shamela.ws/cdn-cgi/challenge-platform/h/b/jsd/r/0.535155722415367:1743200880:Js2GqJ8r0mVeixQ0V59lMnKwflNY_EepznoI1iFADJE/927b02430d5cf851?. 
or 204 http codes., content length is 0? is it safe to ignore these codes? also noticed " preparing redirections showing :1 .  no idea for me about this 1 redirect. entries become very low in numbers when it ends.  overall he has fixed the server. you can proceed the tasks. 

for the artifacts, I'll download them, thank you so much for keeping the warcs. what do you mean with just reprocess them here?  sorry for all these questions, I need just a simple answers, am a little bit new to this.

**Comment by benoit74** [comment]
These logs are not warning or errors. Everything which is at INFO or DEBUG level is mostly plain "normal", it is just here to help understand what is going on in the tool and diagnose potential problem in final result. Even WARNING level could have no consequence. ERROR is general has to be considered as a problem, even if there are exceptions. 

> for the artifacts, I'll download them, thank you so much for keeping the warcs. what do you mean with just reprocess them here?  

We can pass `--warcs xxxx` to zimit, where xxx is the URL or path of of WARC files to process, and in this case zimit will only run the conversion of these WARC files to ZIMs, which is much faster (and less risky in term of different result) than running the crawl to produce these WARCs + converting to ZIM.

**Comment by hamoudak** [comment]
oh. got it. thanks for the info.

**Comment by hamoudak** [comment]
the crawler stopped; {"timestamp":"2025-03-30T22:30:08.632Z","logLevel":"warn","context":"pageStatus","message":"Page Invalid Status: will retry","details":{"retry":0,"retries":2,"url":"https://shamela.ws/book/1050/1614","status":522,"page":"https://shamela.ws/book/1050/1614","workerid":0}}.

the website give, Connection timed out Error code 522.
may be you can start the crawler after it's up.

**Comment by hamoudak** [comment]
why you haven't cancelled this recipe? it has 522 errors, means missing pages will be in the cat . am saying that to gain time to do it from the start or to go for the next one for now.

**Comment by benoit74** [comment]
Are you speaking about https://farm.openzim.org/recipes/shamela.ws_ar_al-buldan-wa-al-rihlat-28 / https://farm.openzim.org/pipeline/11d1f129-c87a-4b79-83b9-57c0e4fabbc7 ?

If yes, then I did not cancelled it because `Page Invalid Status: will retry` means that page will be retried. So you can't know in advance if retry will be successful or not, you have to wait for crawl to complete.

Coping with temporary website failures like this task experienced is exactly the purpose of automatic retries which have been implemented in the crawler by webrecorder team.

Now that crawl + warc2zim is complete, I've checked logs + resulting ZIM.

Log says:

```
{"timestamp":"2025-04-01T09:00:17.667Z","logLevel":"info","context":"worker","message":"Worker done, all tasks complete","details":{"workerid":1}}
{"timestamp":"2025-04-01T09:00:17.823Z","logLevel":"info","context":"worker","message":"Worker done, all tasks complete","details":{"workerid":3}}
{"timestamp":"2025-04-01T09:00:17.894Z","logLevel":"info","context":"worker","message":"Worker done, all tasks complete","details":{"workerid":0}}
{"timestamp":"2025-04-01T09:00:18.123Z","logLevel":"info","context":"worker","message":"Worker done, all tasks complete","details":{"workerid":2}}
{"timestamp":"2025-04-01T09:00:18.602Z","logLevel":"info","context":"crawlStatus","message":"Crawl statistics","details":{"crawled":58636,"total":58636,"pending":0,"failed":0,"limit":{"max":0,"hit":false},"pendingPages":[]}}
{"timestamp":"2025-04-01T09:00:18.605Z","logLevel":"info","context":"general","message":"Crawling done","details":{}}
{"timestamp":"2025-04-01T09:00:18.606Z","logLevel":"info","context":"general","message":"Exiting, Crawl status: done","details":{}}
```

The `"failed":0` at the end of the crawl indicates that all pages have been retried, and they are all successful. 

I sampled about 10-15 pages which had a 522 (they are well indicated in warc2zim logs) and they are indeed all well in the ZIM.

To me, this task completed successfully, we can create ZIM in production library now. Can you help me and check on your own and report any problem you might discover?

**Comment by hamoudak** [comment]
I can confirm most pages that have issues is there in the zim file [shamela.ws_ar_al-buldan-wa-al-rihlat-28]. so the first recipe shamela.ws_ar_al-adab-32 will be the same. I haven't checked its pages till now, checked some, they're also there. I said that it has missing pages according to the http codes. and didn't know the {retries} was implemented. I will double ckeck all. but don't publish https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-tajwid-wa-al-qiraat_2024-12/shamela.ws/category/5. it has missing pages. it was scraped - 2024-12

**Comment by hamoudak** [comment]
@benoit74 there's something wrong with the description of the one you crawling now : 
https://farm.openzim.org/recipes/shamela.ws_ar_al-fiqh-al-hanafi-14/config
description | فقه المذهب المالكي 
it is supposed to be : فقه المذهب الحنفي
the text file I provided you doesn't have the wrong description.

* try to update it the next time.

**Comment by benoit74** [comment]
Indeed, this looks like a copy-paste issue. Fixed.

**Comment by hamoudak** [comment]
I have some questions about a few recipes you have crawled, the first one "shamela.ws_ar_al-adab-32" that you have paused, will you crawl it once again, do you concider it succeded? I have checked the pages that have issues and found out they are all there in the zim file.  I know that Periodicity is manually.  second, for the one that had wrong description, will you crawl it after crawling the recipes in pipline to be updated. lastly, are all recipes will be added to "library.kiwix" or just "dev.library.kiwix".

**Comment by benoit74** [comment]
Current status is:
- all recipes are still in debugging phase, they all publish to dev.library.kiwix.org instead of library.kiwix.org
- only few recipes have been activated in dev (smallest ones) so that we can check their behavior
- I'm still waiting for @Popolechien feedback on recipes which completed in dev to move them to prod
- we've identified an issue with ZIM illustration size, I will have to fix it manually by resizing the illustration (not a big deal at all)

**Comment by Popolechien** [comment]
Looks good to me overall, though I suspect the top link (highlighted below) which normally points to Shamela.ws should point to the selection's default landing page. Same thing for the home link (also highlighted). `Authors` (next to home) is also missing but probably less important in terms of UX.

![Image](https://github.com/user-attachments/assets/0831ae08-82f4-4968-b1c9-62c3c6565f1c)

![Image](https://github.com/user-attachments/assets/102664a7-1922-472f-899a-518040aa9f0b)

**Comment by Popolechien** [comment]
Looking at https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-ilal-wa-al-sualat_2025-04/shamela.ws/author/381 if one trys to click on the third link in the list one gets an error `out of zim` message.

 I know it is in a way expected behaviour, but shouldn't these links rather not be clickable, or maybe a replacement message that says "this content is not part of the current subset, but you can find it in [zim]"?

**Comment by benoit74** [comment]
> Looks good to me overall

Does it means it is ok for you to publish these ZIMs or your points are blocker for insufficient quality? I do not expect to see corresponding features implemented in 2025.

> though I suspect the top link (highlighted below) which normally points to Shamela.ws should point to the selection's default landing page. Same thing for the home link (also highlighted). Authors (next to home) is also missing but probably less important in terms of UX.

Not something feasible with warc2zim currently. Feel free to open a feature request explaining that you want a solution to specify few links which are modified for something totally different, mentioning this use case.

> I know it is in a way expected behaviour, but shouldn't these links rather not be clickable, or maybe a replacement message that says "this content is not part of the current subset, but you can find it in [zim]"?

This is indeed expect behavior, what you would like is not something feasible with warc2zim currently. Feel free to open (another) a feature request explaining how we should detect which links are real external links and which links are to be removed, or how to specify message.

**Comment by Popolechien** [comment]
Yeah so pretty much everything works as expected (or within current, known limitations). I would push to prod but am not familiar enough with the content to decide whether it is usable. I defer to @hamoudak for that.

**Comment by hamoudak** [comment]
@benoit74  you mean resizing favicons, ok.  for the first points that Popolechien talking about, I think this is because the crawl is custom scope or the structure of the website, am not sure. anyway you won't need them in the content when you read the books, but there is something @benoit74  can achieve here, is he crawls the authors page individually, so the reader can have a look on all the authors works. maybe we will have the same issue that for example this sorting https://shamela.ws/authors/?sort=num or https://shamela.ws/authors/?sort=alpha won't work except the default one will do. this is enough for now. so, everything working very well to be pushed to prod. I would also like to add if they're features to be implemented in the future, that will be nice.

**Comment by hamoudak** [comment]
If I downladed or crawled a book with this big size illustration, is it a problem or its just for appropriate looking to the library? I've made many crawls before with it and there was no issue in verbose log.

**Comment by benoit74** [comment]
There is no end-user impact beside 1M "lost" on disk.

**Comment by hamoudak** [comment]
I found out that the domain owner updated the books with PDF version, have you seen it? its' here https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-fiqh-al-hanafi_2025-04/shamela.ws/book/46/78
https://dev.library.kiwix.org/catch/external?source=https%3A%2F%2Fready.shamela.ws%2Fpdf%2Fpdfjs.html%3Ffile%3Dhttps%253A%252F%252Fready.shamela.ws%252Fpdf%252F03-87113%252F01_87111.pdf%23page%3D75

can zimit fetch this pdf version if you include it in anyway? from my experience in crawling, can't fetch this in custom scope. what do you think? if we need these versions with the crawl, any ideas ?

lastly, you'll find the pdf icon of each book online when you browse the book, not in the the main page.  note that not all the library but some books has pdf version for now.
https://shamela.ws/book/1509/46
you'll see it beside search icon.

**Comment by hamoudak** [comment]
some pages in this recipe not found when I check log issues, this happend in the last days, maybe it's temporary. other pages not defined in log is also not found.
DEBUG:Skipping item at shamela.ws/book/149697/4387 whose content length is
 DEBUG:Skipping item at shamela.ws/book/149697/7752 whose content length is 0
https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-fiqh-al-maliki_2025-04/shamela.ws/book/149697/7735
https://dev.library.kiwix.org/viewer#shamela.ws_ar_al-fiqh-al-maliki_2025-04/shamela.ws/book/149697/7752

**Comment by hamoudak** [comment]
@benoit74  Could you please activate this recipe for me shamela.ws_ar_al-adab_2025-03.zim to dev.library, unfortunately, I haven't download it yet, I really was going to buy a flash drive these days to get shamela library. and I think it was a succeeded recipe, beside its importance to my collection.

**Comment by benoit74** [comment]
> can zimit fetch this pdf version if you include it in anyway? from my experience in crawling, can't fetch this in custom scope. what do you think? if we need these versions with the crawl, any ideas ?

Yes we can, just a matter of including the ready.shamela.ws domain inside the scope.

> Could you please activate this recipe for me shamela.ws_ar_al-adab_2025-03.zim to dev.library

Done, but system will first finish ongoing stuff

> some pages in this recipe not found when I check log issues,

I will investigate, probably just restart the recipe

**Comment by hamoudak** [comment]
thank you so much.

**Comment by hamoudak** [comment]
I have included the pdf domain in scope for a small recipe to test this but no success. https://shamela.ws/book/7776
No webpage was found for the web address:

https://pwa.kiwix.org/shamela.ws_ar_sahifat-hammam_2025-04.zim/C/ready.shamela.ws/pdf/pdfjs.html?file=https:/ready.shamela.ws/pdf/waq8910/8910.pdf

**Comment by hamoudak** [comment]
Crawler stopped at 
Scraper progress | 47%
{"crawled":74970,"total":143007,"pending":4,"failed":0,"

it could be a server restart.
shamela.ws_ar_al-adab-32
https://farm.openzim.org/pipeline/239bf8e1-b1ed-4dcc-bb5c-99e7aee05fc0/debug

**Comment by benoit74** [comment]
Thank you, I'm cancelling / restarting the task, it looks like a worker restart indeed.

**Comment by benoit74** [comment]
We have some tasks which completed, are corresponding ZIMs OK and shall I move the corresponding recipes to prod?
- https://farm.openzim.org/pipeline/1314efe4-41b2-49b2-8411-1fa92cbc1edf
- https://farm.openzim.org/pipeline/ab16f84b-ec31-4582-98f3-b72dedb0101e
- https://farm.openzim.org/pipeline/360218bf-7fba-4d50-9aeb-e22ca97a2b7a

**Comment by benoit74** [comment]
One more seems to be ready: https://farm.openzim.org/pipeline/d29bfd7b-8445-449f-97a4-e8628df7dd1e

WDYT?

**Comment by hamoudak** [comment]
Yes, everything is fine.  why you have restarted all three recipes above? is there something wrong with them? and I wanted to tell you something else, when I open kiwix Library i got this https://demo.library.kiwix.org/ 404 Not Found for two days.

**Comment by benoit74** [comment]
I've restarted these recipes to push the file to prod (library.kiwix.org / download.kiwix.org) in a clean way so that they are really available to all our users / officially released. Will do the same for next ones. Next update of these recipes will go straight to prod (but I do not expect to run them very soon).

What do you mean by "open Kiwix Library" (which app / browser, where do you click, what do you type, ...)

**Comment by hamoudak** [comment]
library.kiwix.org , when i open the website with chrome, it redirects me to https://demo.library.kiwix.org/.

<img width="960" alt="Image" src="https://github.com/user-attachments/assets/55e7cb61-7f66-4e2b-b8ab-802b0af0e561" />

**Comment by benoit74** [comment]
Looks like your browser cached something wrong, please clear your browser cache and it will most probably resolve the issue (it works well on my machine and many other ones, and we've indeed made some changes around that recently).

**Comment by hamoudak** [comment]
solved as you suggested.

**Comment by hamoudak** [comment]
there's an issue in uploading artifacts here: https://farm.openzim.org/pipeline/0645e3f6-2db3-49df-a968-e0f66217ace3/debug

**Comment by benoit74** [comment]
Thank you, this has no consequence.

**Comment by hamoudak** [comment]
I can't download it, 
https://s3.us-west-1.wasabisys.com/org-kiwix-zimfarm-artifacts/0645e3f6-2db3-49df-a968-e0f66217ace3_zimit.tar
I get This XML file does not appear to have any style information.

**Comment by hamoudak** [comment]
some recipes are failing now (2), the hardest ones. I would like to scrape them with this version [zimit:3.0.3]. this one was the better for these recipes. 
there was no Link extraction timedout in the logs when the scraper crawls slowly loading pages. it deals very well with the website. https://farm.openzim.org/pipeline/466196d7-aa93-40cd-aec4-d8fb49294255/debug
specially this journal book https://shamela.ws/book/29674

for the last recipe, it failed here https://api.farm.openzim.org/v1/tasks/122748cb-7c81-472b-a530-ab7c4b4f1fc3 .  https://shamela.ws/book/639. this one also loading very slowly.
Browser disconnected (crashed?), interrupting crawl.

**Comment by benoit74** [comment]
Yep, saw this as well ...

**Comment by hamoudak** [comment]
those recipes need to be added to the library
shamela.ws_ar_al-aqida-1
shamela.ws_ar_al-tajwid-wa-al-qiraat-5

you could make these recipes without any issues:

shamela.ws_ar_al-raqaiq-wa-al-adhkar-23
shamela.ws_ar_al-sira-al-nabawiya-24
shamela.ws_ar_al-nahw-wa-al-sarf-31
shamela.ws_ar_al-tib-38
shamela.ws_ar_al-urud-wa-al-qawafi-33
shamela.ws_ar_kutub-amma-39
shamela.ws_ar_ulum-ukhra-40
shamela.ws_ar_al-takhrij-wa-al-atraf-8
shamela.ws_ar_ulum-al-hadith-10
shamela.ws_ar_usul-al-fiqh-11
shamela.ws_ar_al-fiqh-al-hanbali-17
shamela.ws_ar_masail-fiqhiya-19
shamela.ws_ar_al-jawami-36
shamela.ws_ar_al-faharis-wa-al-adilla-37

not sure :
shamela.ws_ar_al-tarajim-wa-al-tabaqat-26
shamela.ws_ar_al-tarikh-25
shamela.ws_ar_shuruh-al-hadith-7

hard recipes: better with  [zimit:3.0.3]

shamela.ws_ar_al-gharib-wa-al-majim-30
for this book - https://shamela.ws/book/29511
shamela.ws_ar_al-sunna-6
shamela.ws_ar_al-fiqh-al-aam-18
for this book - https://shamela.ws/book/11430
[shamela.ws_ar_shuruh-al-hadith-7](https://farm.openzim.org/recipes/shamela.ws_ar_shuruh-al-hadith-7)

In addition to Failure ones:

[shamela.ws_ar_al-adab-32]
[shamela.ws_ar_al-tafsir-3]

Edit: you could temporarily make these hard or failure ones with the latest version but you'll exclude the lazy loading books ids (that caused failure) from the crawling.

**Comment by hamoudak** [comment]
@benoit74 Could you downgrade to [zimit:3.0.3] just for this Library. It'll speed things up.  the"Link extraction timed out" pages is so much with the latest zimit version, you'll have a slow crawling. am sure it will succeed with no issues.

**Comment by hamoudak** [comment]
Have you noticed that "Scraper progress | 100% (2052 / 2052)" stuck since yesterday. I think downgrading is the solution for this issue.

**Comment by hamoudak** [comment]
I think this crawl shamela.ws_ar_ulum-al-fiqh-12 stuck for hours, I have noticed that, it's log full of Page Worker Timeout" then "Page Finished Timed Out", it needs to be resarted.

**Comment by hamoudak** [comment]
@benoit74 hii I have found that the last book of the zim recipe "shamela.ws_ar_ulum-al-fiqh_2025-11" is not in the zim file , it was added lately to the online library on the main page. this one : https://shamela.ws/book/30152 , please note that the author page is there in the zim file, it has just one book. when these books will be updated (the books IDs) ?  also will you scrape the paused recipes of the library? and what about the recipes that failed in zimfarm but uploaded to https://download.kiwix.org/zim/zimit/ ? are they fine to download them.  for example, this one [shamela.ws_ar_al-buldan-wa-al-rihlat-28], [shamela.ws_ar_al-ilal-wa-al-sualat-9] , and [shamela.ws_ar_al-fiqh-al-hanafi-14].

**Comment by benoit74** [comment]
Book is missing because I do not achieve to run the tool which is supposed to update the book in each recipe, looks like shamela.ws has now blacklisted my machine IP. Didn't had time to look into the details.

Didn't had time either to check if we can scrape paused recipes for now.

Recipes which achieved to upload are always supposed to be fine, they currently fail at a later stage (uploading ZIM quality results to the Zimfarm).

**Comment by hamoudak13** [comment]
hey, it's me, the website is now free temporarily from cloudflare message, so the scraper works fine, but maybe it comes once again and so on. In the mean time, there's an issue about the fonts, it needs to be solved, the books font is special and needs to be not changed.
when I've opened the first file : shamela.ws_ar_al-balagha_2026-07, it looked as you see here 

<img width="940" height="476" alt="Image" src="https://github.com/user-attachments/assets/b457540e-95bd-47b5-ac31-2377cc588c1a" />

while in the main website (online) it's :

<img width="956" height="480" alt="Image" src="https://github.com/user-attachments/assets/d5751970-a5e8-49ea-8e79-caf75ed85b51" />

could you please solve this before you continue the crawling process? 
you can delay page crawling time to be away from cloudflare for now.  and you have the chance to run the paused recipes and to use the tool to update the books. also there is an update of the pages online when you browse each page, you see " load the previous page" and "load the next page", If zimit can get these working, it'll be great!. please keep artifacts of each recipe. thank you for your interest. am sorry for being away the last couple months, I was busy at work.

**Comment by hamoudak13** [comment]
I've made a test to crawl a book to see the fonts, it looks fine with zimit:2.1.5 or :2.1.8 . the testing book url is https://shamela.ws/book/23193/2

<img width="953" height="469" alt="Image" src="https://github.com/user-attachments/assets/a5557fce-1229-4096-9b1a-82557c07b9d5" />

**Comment by benoit74** [comment]
I feel like the font issue might be linked to a corrupted font on their website at some point in time. I've requested ZIM again to check if issue is still present ATM.

**Comment by hamoudak13** [comment]
I've tested most of the time, Could you try running a test build completely without the --custom-css argument to see if the default fonts render?

**Comment by benoit74** [comment]
I don't see how the custom CSS option could have an impact. In the ZIM, the font file itself is no a valid font when downloaded. It has no link with HTML, CSS or JS content, the font file should not be broken.

**Comment by hamoudak13** [comment]
you're right, this is a log file of a book I have crawled with zimit:2.1.5, may be help you .

[crawl-20260731003106593.log](https://github.com/user-attachments/files/30722816/crawl-20260731003106593.log)

https://limewire.com/d/OC8Rl#kK5smJaZy2

**Comment by benoit74** [comment]
Can you please have a look at [https://browse.library.kiwix.org/#lang=ara&q=+35+المجموعة](https://browse.library.kiwix.org/#lang=ara&q=+35+%D8%A7%D9%84%D9%85%D8%AC%D9%85%D9%88%D8%B9%D8%A9)?

Looks like https://browse.library.kiwix.org/viewer#shamela.ws_ar_al-balagha_2026-08/shamela.ws/book/12055 is using the same font as https://shamela.ws/book/12055 now ; do you spot other font issues?
